main.tf

```
@@ -214,385 +127,52 @@ resource "aws_s3_bucket" "s3_bucket" {
-
-    tags = var.lifecycle_infrequent_storage_object_tags
-
-    transition {
-      days          = var.lifecycle_days_to_infrequent_storage_transition
-      storage_class = "STANDARD_IA"
-    }
-
-    dynamic "noncurrent_version_transition" {
-      for_each = var.transition_noncurrent_versions == "false" ? [] : [1]
-      content {
-        days          = var.lifecycle_days_to_infrequent_storage_transition
-        storage_class = "STANDARD_IA"
-      }
+      index_document = var.website_index_document
+      error_document = var.website_error_document
     }
   }
 
-  lifecycle_rule {
-    id      = "transition-to-glacier"
-    enabled = var.lifecycle_glacier_transition_enabled
-
-    prefix = var.lifecycle_glacier_object_prefix
-
-    tags = var.lifecycle_glacier_object_tags
-
-    transition {
-      days          = var.lifecycle_days_to_glacier_transition
-      storage_class = "GLACIER"
-    }
-
-    dynamic "noncurrent_version_transition" {
-      for_each = var.transition_noncurrent_versions == "false" ? [] : [1]
-      content {
-        days          = var.lifecycle_days_to_glacier_transition
-        storage_class = "GLACIER"
-      }
-    }
-  }
-
-  lifecycle_rule {
-    id      = "expire-objects"
-    enabled = var.lifecycle_expiration_enabled
-
-    prefix = var.lifecycle_expiration_object_prefix
-
-    tags = var.lifecycle_expiration_object_tags
-
-    expiration {
-      days = var.lifecycle_days_to_expiration
-    }
-
-    dynamic "noncurrent_version_expiration" {
-      for_each = var.expire_noncurrent_versions == "false" ? [] : [1]
-      content {
-        days = var.lifecycle_days_to_expiration
-      }
-    }
-  }
-
-  website {
-    index_document = var.website_index_document
-    error_document = var.website_error_document
-  }
-
   dynamic "server_side_encryption_configuration" {
-    for_each = var.server_side_encryption_configuration
+    for_each = var.website_hosting ? [] : [1]
```

I don't think these should be linked to this website hosting variable. We do have cases where users don't want KMS encryption, but aren't using the website hosting option.

Also, this will be a problem for users who use these buckets as destination buckets for S3 server access logs, but want KMS encryption.  From https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucket-encryption.html:
```
Amazon S3 buckets with default bucket encryption using SSE-KMS cannot be used as destination buckets for Logging requests using server access logging. Only SSE-S3 default encryption is supported for server access log destination buckets.
```

Under the current module, since the encryption is enforced via policies this works so users are forced to encrypt via KMS when putting objects, but log delivery still works.

---

Valid point regarding some tenants not wanting KMS encryption, although I can't think of a use case where SSE-S3 encryption would be preferable since the IAM user will have the right privileges for the KMS key to decrypt the bucket objects. Maybe the cost of decryption. Anyway, I'll make opting out of KMS an option (thus using SSE-S3 instead).

Very strange, about the AWS link above...
I've created a S3 server access logs bucket that is KMS encrypted and it's receiving logs...
See buckets `acp-notprod-tunnel-pnn-tgw-config-audit-eu-west-2` and `acp-notprod-tunnel-pnn-tgw-config-eu-west-2` in notprod

---

> I've created a S3 server access logs bucket that is KMS encrypted and it's receiving logs...
See buckets `acp-notprod-tunnel-pnn-tgw-config-audit-eu-west-2` and `acp-notprod-tunnel-pnn-tgw-config-eu-west-2` in notprod

In the `acp-notprod-tunnel-pnn-tgw-config-audit-eu-west-2` I can see logs up until the 6th August, which based on the commit for that bucket looks like the same day it was changed to use KMS encryption by default instead of AES256. I don't know how often the other bucket gets accessed but there are a lot of logs for the 5th and 6th and then nothing after that, which seems odd to me

---

> I can't think of a use case where SSE-S3 encryption would be preferable since the IAM user will have the right privileges for the KMS key to decrypt the bucket objects

IIRC they're using a tool to do something with the bucket that tool doesn't/didn't support KMS encryption

---

That makes sense. Looks like you found a bug in my setup ;)

Do you think that an extra variable such as `use_default_encryption` would make sense? Defaulting to false so that KMS is used by default, but SSE-S3 is used otherwise (basically for website buckets, audit buckets and buckets for which tenants don't want to use KMS).

---

An extra variable sounds good to me. Maybe something more like `use_kms_encryption`? `use_default_encryption` sounds a little misleading IMO (e.g. if I set that to false without knowing exactly what it did, I'd expect that the bucket to not have encyption)

---

As discussed over Slac, I used the existing `km_alias` variable to drive the logic.
Basically, KMS is used if and only if `website_hosting != true` and `kms_alias != ""`